cardano-api/src/Cardano/Api/Value.hs

```diff
@@ -366,11 +360,28 @@ instance FromJSON ValueNestedRep where
     where
       parsePid :: (Text, Aeson.Value) -> Parser ValueNestedBundle
       parsePid ("lovelace", q) = ValueNestedBundleAda <$> parseJSON q
-      parsePid (pid, q) =
-        case deserialiseFromRawBytesHex AsScriptHash (Text.encodeUtf8 pid) of
-          Just sHash -> ValueNestedBundle (PolicyId sHash) <$> parseJSON q
-          Nothing -> fail $ "Failure when deserialising PolicyId: "
-                         <> Text.unpack pid
+      parsePid (pid, quantityBundleJson) = do
+        sHash <-
+          note ("Failure when deserialising PolicyId: " <> Text.unpack pid) $
+          deserialiseFromRawBytesHex AsScriptHash $ Text.encodeUtf8 pid
+        quantityBundleRaw <- parseJSON quantityBundleJson
+        quantityBundle <-
+          Map.fromList
+          <$> for (Map.assocs quantityBundleRaw) \(nameText, quantity) -> do
+                name <-
+                  note
+                    ("Failure when deserialising AssetName"
+                      <> Text.unpack nameText) $
+                  deserialiseFromRawBytesHex AsAssetName $
+                  Text.encodeUtf8 nameText
+                pure (name, quantity)
+        pure $ ValueNestedBundle (PolicyId sHash) quantityBundle
+
+note :: MonadFail m => String -> Maybe a -> m a
```

Comment:
`failWith` or something similar would be a better name.

Reply:
this name is used in Cardano.Prelude (but with a slightly different type) and https://hackage.haskell.org/package/errors-2.3.0/docs/Control-Error-Util.html#v:note. I consider it somewhat traditional.

Reply:
`note` needs to be removed from here